lib/Service/BlockAgentService.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+declare(strict_types=1);
+/**
+ * @copyright Copyright (c) 2021, Vitor Mattos <vitor@php.rio>
+ *
+ * @author Vitor Mattos <vitor@php.rio>
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Talk\Service;
+
+use OCA\Talk\Model\BlockAgent;
+use OCA\Talk\Model\BlockAgentMapper;
+use OCP\AppFramework\Utility\ITimeFactory;
+use OCP\DB\Exception;
+use OCP\ICache;
+use OCP\ICacheFactory;
+
+class BlockAgentService {
+
+	/** @var BlockAgentMapper */
+	private $blockAgentMapper;
+	/** @var ITimeFactory */
+	private $timeFactory;
+	/** @var ICache */
+	private $cache;
+
+	public function __construct(BlockAgentMapper $blockAgentMapper,
+								ITimeFactory $timeFactory,
+								ICacheFactory $cacheFactory) {
+		$this->blockAgentMapper = $blockAgentMapper;
+		$this->timeFactory = $timeFactory;
+		$this->cache = $cacheFactory->createDistributed('talk_blocked_users');
+	}
+
+	public function block(string $actorType, string $actorId, string $blockedType, string $blockedId) {
+		$blockAgent = new BlockAgent();
+		$blockAgent->setActorType($actorType);
+		$blockAgent->setActorId($actorId);
+		$blockAgent->setBlockedType($blockedType);
+		$blockAgent->setBlockedId($blockedId);
+		$blockAgent->setDatetime($this->timeFactory->getDateTime());
+		try {
+			$this->blockAgentMapper->insert($blockAgent);
+			$blockedList = $this->cache->get($actorId) ?? [];
+			$blockedList[] = $blockAgent;
+			$this->cache->set($actorId, $blockedList);
```

Comment:
This is assuming only ever users would block, but it might be a guest feature too at some point.
I would also not cache it for now, until we know where we use it and which caching layout makes most sense.

Reply:
Cache removed at this time